ghedesigner/main.py

```diff
@@ -0,0 +1,272 @@
+import logging
+from json import loads
+from pathlib import Path
+from sys import exit, stderr
+
+import click
+from jsonschema import ValidationError
+
+from ghedesigner import VERSION
+from ghedesigner.enums import BHPipeType, DesignGeomType
+from ghedesigner.ghe.manager import GroundHeatExchanger
+from ghedesigner.heat_pump import HeatPump
+from ghedesigner.utilities import write_idf
+from ghedesigner.validate import validate_input_file
+
+logging.basicConfig(level=logging.WARN, format="%(message)s", datefmt="[%X]")
+logger = logging.getLogger(__name__)
+
+
+def _run_manager_from_cli_worker(input_file_path: Path, output_directory: Path) -> int:
+    """
+    Worker function to run simulation.
+
+    :param input_file_path: path to input file. Input file must exist.
+    :param output_directory: path to write output files. Output directory must be a valid path.
+    """
+
+    # validate inputs against schema before doing anything
+    if validate_input_file(input_file_path) != 0:
+        return 1
+
+    inputs = loads(input_file_path.read_text())
+
+    # Read in all the inputs into small dicts
+    # it is possible to define multiple fluids, GHEs, and boreholes in the inputs, I'm just taking the first for now
+    input_file_version: int = inputs["version"]
+    if input_file_version != 1:
+        print("Bad input file version, right now we support these versions: 1")
+        return 1
+
+    # Process the load source, it should be a building object or a GHE with loads specified
+    ghe_contains_loads = ["loads" in ghe_dict for _, ghe_dict in inputs["ground-heat-exchanger"].items()]
+    all_ghe_has_loads = all(ghe_contains_loads)
+    no_ghe_has_loads = all(not x for x in ghe_contains_loads)
+    building_input = "building" in inputs
+    valid_load_source = all_ghe_has_loads ^ (building_input and no_ghe_has_loads)  # XOR because we don't want both
+    if not valid_load_source:
+        print("Bad load specified, need exactly one of: loads in each ghe, or building object")
+        return 1
+
+    # Get a couple required objects from the inputs
+    sim_control = inputs["simulation-control"]
+    fluid_props = inputs["fluid"]
+
+    # Loop over the topology and init the found objects, for now just the GHE and an optional building
+    topology_props: list[dict] = inputs["topology"]
+    heat_pump: HeatPump
+    ghe = GroundHeatExchanger()
+    for component in topology_props:
```

Comment:
Everything is now guided by the high level topology object.  This should allow us to do more flexible things with the code.

Reply:
Love it.